plugins/tickets.py

```diff
@@ -0,0 +1,1821 @@
+        roleid BIGINT PRIMARY KEY
+);
+
+CREATE TABLE tickets.history (
+        version INT,
+        last_modified_at TIMESTAMP,
+        id INT,
+        type INT,
+        stage INT,
+        status INT,
+        modid BIGINT,
+        targetid BIGINT,
+        roleid BIGINT,
+        auditid BIGINT,
+        duration INT,
+        comment TEXT,
+        list_msgid BIGINT,
+        delivered_id BIGINT,
+        created_at TIMESTAMP,
+        modified_by BIGINT,
+        PRIMARY KEY (id, version),
+        FOREIGN KEY (id) REFERENCES tickets.tickets ON UPDATE CASCADE
+);
+
+CREATE FUNCTION tickets.log_ticket_update() RETURNS TRIGGER AS $log_ticket_update$
+        DECLARE
+                modified tickets.tickets%rowtype;
+                last_version int;
+        BEGIN
+                SELECT INTO modified
+                        NEW.id,
+                        NULLIF(NEW.type, OLD.type),
+                        NULLIF(NEW.stage, OLD.stage),
+                        NULLIF(NEW.status, OLD.status),
+                        NULLIF(NEW.modid, OLD.modid),
+                        NULLIF(NEW.targetid, OLD.targetid),
+                        NULLIF(NEW.roleid, OLD.roleid),
+                        NULLIF(NEW.auditid, OLD.auditid),
+                        NULLIF(NEW.duration, OLD.duration),
+                        NULLIF(NEW.comment, OLD.comment),
+                        NULLIF(NEW.list_msgid, OLD.list_msgid),
+                        NULLIF(NEW.delivered_id, OLD.delivered_id),
+                        NULLIF(NEW.created_at, OLD.created_at),
+                        NEW.modified_by;
+                SELECT version INTO last_version FROM tickets.history WHERE id = OLD.id ORDER BY version DESC LIMIT 1;
+                IF NOT FOUND THEN
+                        INSERT into tickets.history VALUES (0, OLD.created_at, OLD.*), (1, now(), modified.*);
+                ELSE
+                        INSERT into tickets.history VALUES (coalesce(last_version + 1, 1), now(), modified.*);
+                END IF;
+                RETURN NULL;
+        END
+$log_ticket_update$ LANGUAGE plpgsql;
+
+CREATE TRIGGER log_update
+        AFTER UPDATE ON tickets.tickets
+        FOR EACH ROW
+        WHEN (OLD.* IS DISTINCT FROM NEW.*)
+        EXECUTE PROCEDURE tickets.log_ticket_update();
+"""
+
+ticket_comment_re = re.compile(
+    r"""
+    (?i)\s*
+    ([\d.]+)
+    (s(?:ec(?:ond)?s?)?
+    |(?-i:m)|min(?:ute)?s?
+    |h(?:(?:ou)?rs?)?
+    |d(?:ays?)?
+    |w(?:(?:ee)?ks?)
+    |(?-i:M)|months?
+    |y(?:(?:ea)?rs?)?
+    )
+    |p(?:erm(?:anent)?)?\W+
+    """.replace(' ', '').replace('\n', '')
+)
+
+time_expansion = {
+    's': 1,
+    'm': 60,
+    'h': 60 * 60,
+    'd': 60 * 60 * 24,
+    'w': 60 * 60 * 24 * 7,
+    'M': 60 * 60 * 24 * 30,
+    'y': 60 * 60 * 24 * 365
+}
+
+# ----------- Config -----------
+conf = db.kv.Config(__name__)  # General plugin configuration
+
+conf.guild: int  # ID of the guild the ticket system is managing
```

Comment:
Unfortunately we cannot store snowflakes in JSON because jq doesn't understand big integers and will truncate to floating precision (snowflakes are large enough that this matters). Consider storing as string.